compiler/injectdestructors.nim

```diff
@@ -390,20 +347,21 @@ proc getTemp(c: var Con; typ: PType; info: TLineInfo): PNode =
   result = newSymNode(sym)
   c.addTopVar(result)
 
-proc p(n: PNode; c: var Con): PNode
+proc pStmt(n: PNode; c: var Con): PNode
+proc pExpr(n: PNode; c: var Con): PNode
 
-template recurse(n, dest) =
-  for i in 0..<n.len:
-    dest.add p(n[i], c)
+template isExpression(n: PNode): bool =
+  (not isEmptyType(n.typ)) or (n.kind in nkLiterals + {nkNilLit})
 
-proc genMagicCall(n: PNode; c: var Con; magicname: string; m: TMagic): PNode =
-  result = newNodeI(nkCall, n.info)
-  result.add(newSymNode(createMagic(c.graph, magicname, m)))
-  result.add n
+template recurse(n, dest) =
+  for son in n:
+    dest.add if isExpression(son): pExpr(son, c)
```

Comment:
This looks sloppy to me. You should know from the context whether to call ``pExpr`` or ``pStmt``. There should be 2 different ``recurse`` templates for this then.

Reply:
Yeah, I agree.